### PR TITLE
[GEOT-5968] Add NativeFilter filter

### PR DESCRIPTION
**Pull request by nmco**
Associated issue:
https://osgeo-org.atlassian.net/browse/GEOT-5968

Associated mailing list:
https://sourceforge.net/p/geotools/mailman/message/36268473/

### CODE REVIEW

File: modules/library/opengis/src/main/java/org/opengis/filter/NativeFilter.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ *    version 2.1 of the License.
+ *
+ *    This library is distributed in the hope that it will be useful,
+ *    but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ *    Lesser General Public License for more details.
+ */
+package org.opengis.filter;
+
+/** */
```
Comment (aaime) [inline modules/library/opengis/src/main/java/org/opengis/filter/NativeFilter.java]: No javadoc? (here and in the method)

File: modules/library/xml/src/main/java/org/geotools/xml/filter/FilterCompliancePreProcessor.java [hunk 2]
```diff
@@ -855,4 +856,9 @@ public String toString() {
     public boolean requiresPostProcessing() {
         return requiresPostProcessing;
     }
+
+    @Override
+    public Object visit(NativeFilter nativeFilter, Object extraData) {
+        throw new UnsupportedOperationException("Native filter is not supported");
```
Comment (aaime) [inline modules/library/xml/src/main/java/org/geotools/xml/filter/FilterCompliancePreProcessor.java]: Seems quite similar to what the default method already offers? But maybe best to keep it, but throw an exception stating that native filters are not encodable in XML (this is what the visitor seem to be doing, prepare filters for XML encoding).

File: modules/plugin/jdbc/jdbc-oracle/src/main/java/org/geotools/data/oracle/OracleFilterToSQL.java [hunk 3]
```diff
@@ -173,6 +174,9 @@ protected FilterCapabilities createFilterCapabilities() {
         caps.addType(EndedBy.class);
         caps.addType(TEquals.class);
 
+        // native filter support
+        caps.addType(NativeFilter.class);
```
Comment (aaime) [inline modules/plugin/jdbc/jdbc-oracle/src/main/java/org/geotools/data/oracle/OracleFilterToSQL.java]: If all the native filter handling in SQL does is to take it and write it out in the SQL, why not add support for it in SQLDialect#BASE_DBMS_CAPABILITIES instead? And selected DBMS can remove it, if needs be (don't think there will be such need)

File: modules/library/jdbc/src/test/java/org/geotools/jdbc/JDBCNativeFilterOnlineTest.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+            features.add(iterator.next());
+        }
+        iterator.close();
+        // check that we retrieved the necessary features
+        assertThat(features.size(), is(1));
+        assertThat(features.get(0).getAttribute(tname("code")), is("#2"));
+    }
+
+    /**
+     * Builds a native filter that consider only the measurement of type temperature and wind with a
```
Comment (aaime) [inline modules/library/jdbc/src/test/java/org/geotools/jdbc/JDBCNativeFilterOnlineTest.java]: I would reword as "measurement of type temperature or type wind, and with a value greater than 15" and add the example SQL too in the javadoc for less ambiguity: "(TYPE = 'temperature' OR TYPE = 'wind') AND value > 15"

File: modules/library/opengis/src/main/java/org/opengis/filter/NativeFilter.java (hunk 1 above)
Comment (nmco) [reply modules/library/opengis/src/main/java/org/opengis/filter/NativeFilter.java]: Ooops, javadoc added.

File: modules/plugin/jdbc/jdbc-oracle/src/main/java/org/geotools/data/oracle/OracleFilterToSQL.java (hunk 3 above)
Comment (nmco) [reply modules/plugin/jdbc/jdbc-oracle/src/main/java/org/geotools/data/oracle/OracleFilterToSQL.java]: Good point, done.

**Comment by nmco** [comment]
Thank you for the detailed review, suggestions applied.
